### PR TITLE
Add ability to download and watch replays

### PR DESCRIPTION
**Pull request by naoey**
Adds a button in the score information section that can download and play back replays.

image

Again, unsure of the UI direction of this but I gave it a shot.

Partially addresses #4991. Alternate implementation of #4985 as discussed.

### CODE REVIEW

File: osu.Game/Screens/Play/ReplayDownloadButton.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        private const int size = 40;
+
+        [Resolved(canBeNull: true)]
+        private OsuGame game { get; set; }
+
+        [Resolved]
+        private ScoreManager scores { get; set; }
+
+        [Resolved]
+        private OsuColour colours { get; set; }
```
Comment (Altenhh) [inline osu.Game/Screens/Play/ReplayDownloadButton.cs]: We could move both `OsuGame game` and `OsuColour colours` into the dependency loader as they are not used anywhere else in the code.

File: osu.Game/Screens/Play/ReplayDownloadButton.cs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+using osu.Framework.Graphics.Effects;
+using osuTK.Graphics;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Game.Screens.Play
+{
+    public class ReplayDownloadButton : DownloadTrackingComposite<ScoreInfo, ScoreManager>, IHasTooltip
+    {
+        private const int size = 40;
```
Comment (Altenhh) [inline osu.Game/Screens/Play/ReplayDownloadButton.cs]: Is there any reason why this is held inside a variable?

**Review by naoei — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Code looks fine.

As discussed in discord, let's change the button to match (at least visually, or potentially code sharing) the direct download panel button:

image

**Comment by naoey** [comment]
Updated to share download button UI between direct panels and score screens:
image

Unsure if the shake container in the new code should be part of the button itself?

File: osu.Game/Overlays/Direct/DownloadButton.cs [hunk 3]
```diff
@@ -105,32 +72,11 @@ private void load(OsuColour colours, OsuGame game, BeatmapManager beatmaps)
...
+        protected override void Dispose(bool isDisposing)
         {
-            switch (state)
-            {
-                case DownloadState.NotDownloaded:
-                    background.FadeColour(colours.Gray4, 500, Easing.InOutExpo);
-                    icon.MoveToX(0, 500, Easing.InOutExpo);
-                    checkmark.ScaleTo(Vector2.Zero, 500, Easing.InOutExpo);
-                    break;
+            base.Dispose(isDisposing);
```
Comment (peppy) [inline osu.Game/Overlays/Direct/DownloadButton.cs]: any specific reason for this? it should be automatically unbound by the child.

Comment (naoey) [reply osu.Game/Overlays/Direct/DownloadButton.cs]: oh, i misunderstood bindings then. will remove.

**Review by peppy — APPROVED** [review]
looks good

**Pull request merged**